scripts/util/charmhub.py

```diff
@@ -156,29 +162,38 @@ def get_charmhub_auth_macaroon() -> str:
         raise ValueError("Malformed charmhub credentials")
     return v
 
+
+def find_revision(charm_name: str, channel: str, arch: str, base: str) -> int | None:
+    log.info(
+        f"Querying Charmhub to get revisions of {charm_name=} {channel=} {arch=} {base=} ..."
+    )
+    url = "https://api.charmhub.io/v2/charms/refresh"
+    headers = {"Content-Type": "application/json"}
+    data = {
+        "actions": [
+            {
+                "action": "install",
+                "base": {"architecture": arch, "channel": base, "name": "ubuntu"},
+                "channel": channel,
+                "name": charm_name,
+                "instance-key": "query",
+            }
+        ],
+        "context": [],
+    }
+    r = requests.post(url, headers=headers, json=data, timeout=TIMEOUT)
+    return r.json()["results"][0]["charm"].get("revision") if r.status_code == 200 else None
+
+
 def get_revision_matrix(charm_name: str, channel: str) -> RevisionMatrix:
     """Get the revision of a charm in a channel."""
-    headers = {
-        "Content-Type": "application/json",
-    }
     log.info(f"Querying Charmhub to get revisions of {charm_name} in {channel}...")
-    
-    url = f"https://api.charmhub.io/v2/charms/info/{charm_name}?fields=channel-map"
-    r = requests.get(url, headers=headers, timeout=TIMEOUT)
-    r.raise_for_status()
-
-    data = json.loads(r.text)
-    log.info("Search for charm revisions in channel list...")
 
     revision_matrix = RevisionMatrix()
-    for channel_map in data.get("channel-map", []):
-        
-        if channel == channel_map["channel"]["name"]:
-            revision_matrix.set(
-                channel_map["channel"]["base"]["architecture"],
-                channel_map["channel"]["base"]["channel"],
-                int(channel_map["revision"]["revision"]),
-            )
+    for base in ["20.04", "22.04", "24.04", "26.04", "28.04", "30.04"]:
+        for arch in ["amd64", "arm64"]:
```

Comment:
using the `refresh` api allows for charm branches to be searched (for instance, to get the revision of a charm in `1.32/edge/etcd-3.6`, i had to use this because branches don't show up in the `info` api